cpp/src/arrow/compute/cast.cc

```diff
@@ -94,7 +94,9 @@ class CastMetaFunction : public MetaFunction {
                             const FunctionOptions* options,
                             ExecContext* ctx) const override {
     ARROW_ASSIGN_OR_RAISE(auto cast_options, ValidateOptions(options));
-    if (args[0].type()->Equals(*cast_options->to_type)) {
+    // args[0].type() could be a nullptr so check for that before
+    // we do anything with it.
+    if (args[0].type() && args[0].type()->Equals(*cast_options->to_type)) {
```

Comment:
Should we ensure the argument is always an Array? Something like:
```
  switch (args[0].kind()) {
    case Datum::ARRAY:
      // Implementation
      break;
    default:
      return Status::NotImplemented("Cast arguments should be array-like");
```

Reply:
I think it should be done to all functions which don't support this. Probably worth an issue by itself?